packages/components/src/focal-point-picker/index.js

```diff
@@ -23,317 +26,170 @@ import {
 	MediaWrapper,
 	MediaContainer,
 } from './styles/focal-point-picker-style';
-import { roundClamp } from '../utils/math';
 import { INITIAL_BOUNDS } from './utils';
 
-export class FocalPointPicker extends Component {
-	constructor( props ) {
-		super( ...arguments );
-
-		this.state = {
-			isDragging: false,
-			bounds: INITIAL_BOUNDS,
-			percentages: props.value,
-		};
-
-		this.containerRef = createRef();
-		this.mediaRef = createRef();
-
-		this.onMouseDown = this.startDrag.bind( this );
-		this.onMouseUp = this.stopDrag.bind( this );
-		this.onKeyDown = this.onKeyDown.bind( this );
-		this.onMouseMove = this.doDrag.bind( this );
-		this.ifDraggingStop = () => {
-			if ( this.state.isDragging ) {
-				this.stopDrag();
-			}
-		};
-		this.onChangeAtControls = ( value ) => {
-			this.updateValue( value, () => {
-				this.props.onChange( this.state.percentages );
-			} );
-		};
-
-		this.updateBounds = this.updateBounds.bind( this );
-		this.updateValue = this.updateValue.bind( this );
-	}
-	componentDidMount() {
-		const { defaultView } = this.containerRef.current.ownerDocument;
-		defaultView.addEventListener( 'resize', this.updateBounds );
-
-		/*
-		 * Set initial bound values.
-		 *
-		 * This is necessary for Safari:
-		 * https://github.com/WordPress/gutenberg/issues/25814
-		 */
-		this.updateBounds();
-	}
-	componentDidUpdate( prevProps ) {
-		if ( prevProps.url !== this.props.url ) {
-			this.ifDraggingStop();
-		}
-		/*
-		 * Handles cases where the incoming value changes.
-		 * An example is the values resetting based on an UNDO action.
-		 */
-		const {
-			isDragging,
-			percentages: { x, y },
-		} = this.state;
-		const { value } = this.props;
-		if ( ! isDragging && ( value.x !== x || value.y !== y ) ) {
-			this.setState( { percentages: this.props.value } );
-		}
-	}
```

Comment:
Are these no longer necessary?

Reply:
It's good to cover this. I'm glad you asked. The near-equivalent logic is here:
https://github.com/WordPress/gutenberg/blob/4a9042fde61f5f33cb72dc894f0e3092d546953b/packages/components/src/focal-point-picker/index.js#L60-L61
The internal state is used only while having not yet propagated (i.e. while dragging) and otherwise the incoming `value` prop is used (i.e. it’s controlled so things like UNDO/REDO will work).

I did decide the logic to stop dragging when the `url` prop changes was unnecessary based on testing and finding it to work fine (I can only imagine this happening due to UNDO/REDO and seems a quite obscure edge case).

I had tested all this months ago in the post editor and now to be sure I put together a story and retested.

<details><summary>Story for controllable <code>FocalPointPicker</code> using <kbd>Meta</kbd>+<kbd>x</kbd> to make controlled swaps media and <kbd>Meta</kbd>+<kbd>z</kbd> to make controlled updates to the value</summary>

```diff
diff --git a/packages/components/src/focal-point-picker/stories/index.js b/packages/components/src/focal-point-picker/stories/index.js
index 475029611a..eec67b82a2 100644
--- a/packages/components/src/focal-point-picker/stories/index.js
+++ b/packages/components/src/focal-point-picker/stories/index.js
@@ -1,7 +1,7 @@
 /**
  * WordPress dependencies
  */
-import { useState } from '@wordpress/element';
+import { useEffect, useRef, useState } from '@wordpress/element';
 /**
  * Internal dependencies
  */
@@ -31,6 +31,49 @@ export const _default = () => {
 	return <Example />;
 };
 
+const Controllable = ( { x = 0.5, y = 0.5 } ) => {
+	const [ focalPoint, setFocalPoint ] = useState( { x, y } );
+	const [ media, setMedia ] = useState( 0 );
+	const refMediaList = useRef( [
+		`data:image/svg+xml,%3Csvg width='280' height='210' xmlns='http://www.w3.org/2000/svg'/%3E`,
+		`data:image/svg+xml,%3Csvg width='480' height='360' xmlns='http://www.w3.org/2000/svg'/%3E`,
+	] );
+	const refHeldKeySet = useRef( new Set() );
+	useEffect( () => {
+		const heldKeySet = refHeldKeySet.current;
+		const downer = ( { key } ) => {
+			heldKeySet.add( key );
+			if ( heldKeySet.has( 'Meta' ) ) {
+				if ( key === 'x' )
+					setMedia(
+						( v ) => ( v + 1 ) % refMediaList.current.length
+					);
+				else if ( key === 'z' )
+					setFocalPoint( ( v ) => ( {
+						x: 1 + v.x * -1,
+						y: 1 + v.y * -1,
+					} ) );
+			}
+		};
+		const upper = ( { key } ) => refHeldKeySet.current.delete( key );
+		document.addEventListener( 'keydown', downer );
+		document.addEventListener( 'keyup', upper );
+		return () => {
+			document.removeEventListener( 'keydown', downer );
+			document.removeEventListener( 'keyup', upper );
+		};
+	}, [] );
+	return (
+		<FocalPointPicker
+			value={ focalPoint }
+			onChange={ setFocalPoint }
+			url={ refMediaList.current[ media ] }
+		/>
+	);
+};
+
+export const controllable = () => <Controllable />;
+
 export const image = () => {
 	const url =
 		'https://i0.wp.com/themes.svn.wordpress.org/twentytwenty/1.3/screenshot.png?w=572&strip=al';

```
</details>

Here is a demo of controlled updates switching the media while dragging and afterward changing the value.

https://user-images.githubusercontent.com/9000376/181604876-9dab9da2-0c22-4cc5-8a88-8cead98677a8.mp4

Reply:
Tested the story, works as expected 👌